File: android/src/main/java/co/alephnull/reactnative/silicon/ReactNativeSiliconModule.kt
package co.alephnull.reactnative.silicon

import co.alephnull.reactnative.silicon.device.SiliconDevice
import co.alephnull.reactnative.silicon.helpers.SiliconHelpers
import co.alephnull.reactnative.silicon.jose.SiliconJose
import co.alephnull.reactnative.silicon.keystoremanager.AttestFormat
import co.alephnull.reactnative.silicon.keystoremanager.GenerateKeyOptions
import co.alephnull.reactnative.silicon.keystoremanager.PubkeyFormat
import co.alephnull.reactnative.silicon.keystoremanager.SiliconKSM
import co.alephnull.reactnative.silicon.randomgen.RandomBytesFormat
import co.alephnull.reactnative.silicon.randomgen.SiliconRandomGen
import co.alephnull.reactnative.silicon.signer.SignDigest
import co.alephnull.reactnative.silicon.signer.SignOptions
import co.alephnull.reactnative.silicon.signer.SiliconSigner
import co.alephnull.reactnative.silicon.verifier.SiliconVerifier
import co.alephnull.reactnative.silicon.verifier.BridgeVerifyOptions
import co.alephnull.reactnative.silicon.verifier.VerifyOptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.security.KeyStore

class ReactNativeSiliconModule : Module() {
    private val keystore by lazy {
        // Instantiate the keystore
        // We pass in "AndroidKeyStore", and pass 'null' into load (Android OS manages the hardware-level storage automatically)
        KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
    }

    private val helpers by lazy { SiliconHelpers() }
    private val device by lazy { SiliconDevice(appContext) }
    private val keystoreManager by lazy { SiliconKSM(appContext, keystore, helpers) }
    private val signer by lazy { SiliconSigner(appContext, keystore, helpers) }
    private val verifier by lazy { SiliconVerifier(keystore) }
    private val randomGen by lazy { SiliconRandomGen() }
    private val jose by lazy { SiliconJose(keystore, helpers) }

    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    override fun definition() = ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('ReactNativeSilicon')` in JavaScript.
        Name("ReactNativeSilicon")

        AsyncFunction("getCapabilities") {
            try {
                val result = device.getCapabilities()
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.GET_CAPABILITIES_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        // ---- Keystore Manager ----

        AsyncFunction("generateKey") { alias: String, opts: GenerateKeyOptions, attestChallenge: ByteArray ->
            try {
                // Uint8Array (TS) fails to map to ByteArray correctly when it is nested inside the options
                // so we extract it out of the options on the TS side and then inject it back in here
                opts.attestChallenge = attestChallenge

                val result = keystoreManager.generateKey(alias, opts)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.GENERATE_KEY_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        AsyncFunction("deleteKey") { alias: String ->
            try {
                val result = keystoreManager.deleteKey(alias)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.DELETE_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        AsyncFunction("deleteAllKeys") { prefix: String? ->
            try {
                val result = keystoreManager.deleteAllKeys(prefix)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.DELETE_ALL_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        AsyncFunction("keyExists") { alias: String ->
            try {
                val result = keystoreManager.keyExists(alias)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.KEY_EXISTS_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        AsyncFunction("listKeys") { prefix: String? ->
            try {
                val result = keystoreManager.listKeys(prefix)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.LIST_KEYS_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        AsyncFunction("getPubKey") { alias: String, format: PubkeyFormat ->
            try {
                val result = keystoreManager.getPubKey(alias, format)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.GET_PUB_KEY_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        // We ignore pubKeyFormat (ios only)
        AsyncFunction("attestKey") { alias: String, format: AttestFormat, pubKeyFormat: String ->
            try {
                val result = keystoreManager.attestKey(alias, format)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.ATTEST_KEY_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        AsyncFunction("getKeyInfo") { alias: String ->
            try {
                val result = keystoreManager.getKeyInfo(alias)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.GET_KEY_INFO_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        AsyncFunction("validateKey") { alias: String ->
            try {
                val result = keystoreManager.validateKey(alias)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.VALIDATE_KEY_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        // ---- Sign/Verify ----

        AsyncFunction("sign") Coroutine { alias: String, payloadStr: String?, payloadByteArr: ByteArray?, opts: SignOptions ->
            try {
                val bridgePayload = BridgePayloadRecord()
                bridgePayload.text = payloadStr
                bridgePayload.bytes = payloadByteArr

                val result = signer.sign(alias, bridgePayload.toPayloadType(), opts)
                return@Coroutine result.toBridgeMap()
            } catch (e: Exception) {
                return@Coroutine SiliconResult.Failure(
                    SiliconErrorCode.SIGN_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        AsyncFunction("verify") {
            payloadStr: String?,
            payloadByteArr: ByteArray?,
            signatureStr: String?,
            signatureByteArr: ByteArray?,
            pubkeyStr: String?,
            pubkeyByteArr: ByteArray?,
            bridgeOpts: BridgeVerifyOptions ->

            try {
                // Pack payload into PayloadType
                val bridgePayload = BridgePayloadRecord()
                bridgePayload.text = payloadStr
                bridgePayload.bytes = payloadByteArr

                // Pack signature into PayloadType
                val bridgeSignature = BridgePayloadRecord()
                bridgeSignature.text = signatureStr
                bridgeSignature.bytes = signatureByteArr

                // Pack pubkeyStr and pubkeyBytes into PayloadType
                var pubkey: BridgePayloadRecord? = null
                if (pubkeyStr != null || pubkeyByteArr != null) {
                    pubkey = BridgePayloadRecord()
                    pubkey.text = pubkeyStr
                    pubkey.bytes = pubkeyByteArr
                }

                // Populate VerifyOptions
                val opts = VerifyOptions()
                opts.pubkey = when (pubkey) {
                    null -> null
                    else -> pubkey.toPayloadType()
                }
                opts.alias = bridgeOpts.alias
                opts.algorithm = bridgeOpts.algorithm

                val result = verifier.verify(
                    bridgePayload.toPayloadType(),
                    bridgeSignature.toPayloadType(),
                    opts
                )
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.VERIFY_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        // ---- Random Generator ----

        AsyncFunction("generateSecureRandomBytes") { length: Int, format: RandomBytesFormat ->
            try {
                val result = randomGen.generate(length, format)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.RANDOM_GEN_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }

        // ---- JOSE ----

        AsyncFunction("getJwk") { alias: String, digest: SignDigest? ->
            try {
                val result = jose.getJwk(alias, digest)
                return@AsyncFunction result.toBridgeMap()
            } catch (e: Exception) {
                return@AsyncFunction SiliconResult.Failure(
                    SiliconErrorCode.GET_JWK_FAILED,
                    e.localizedMessage ?: "Failed to due to unexpected error"
                ).toBridgeMap()
            }
        }
    }
}

File: ios/ReactNativeSiliconModule.swift
import ExpoModulesCore

public class ReactNativeSiliconModule: Module {
    // MARK: - Providers (External dependencies)
    private lazy var attestService: AttestServiceProvider = AttestService()
    private lazy var secItems: SecItemsProvider = SecItems()
    private lazy var secKeys: SecKeysProvider = SecKeys()
    private lazy var secAccessControls: SecAccessControlsProvider = SecAccessControls()
    private lazy var secRandom: SecRandomProvider = SecRandom()
    private lazy var secureEnclave: SecureEnclaveProvider = SecureEnclave()
    private lazy var authContextSession: AuthContextSessionProvider = AuthContextSession()
    
    // MARK: - Shared (Shared core functionality)
    private lazy var authContextCache: AuthContextCache = AuthContextCache(authContextSession: authContextSession)
    private lazy var keychainHelper: KeychainHelping = KeychainHelper(secItems: secItems)
    private lazy var keyMetadataStore: KeyMetadataStoring = KeyMetadataStore(keychainHelper: keychainHelper)
    private lazy var pubKeyData: PubKeyData = PubKeyData(secItems: secItems, secKeys: secKeys)
    
    // MARK: - Handlers (Handling bridge requests)
    private lazy var siliconDevice: SiliconDevice = SiliconDevice(
        authContextSession: authContextSession,
        secureEnclave: secureEnclave,
        attestService: attestService
    )
    
    private lazy var keyGenerator: KeyGenerating = KeyGenerator(
        attestService: attestService,
        secItems: secItems,
        secKeys: secKeys,
        secAccessControls: secAccessControls,
        keychainHelper: keychainHelper
    )
    private lazy var siliconKSM = SiliconKeystoreManager(
        attestService: attestService,
        secItems: secItems,
        secAccessControls: secAccessControls,
        secureEnclave: secureEnclave,
        authContextSession: authContextSession,
        pubKeyData: pubKeyData,
        keyGenerator: keyGenerator,
        keychainHelper: keychainHelper,
        keyMetadataStore: keyMetadataStore
    )
    
    private lazy var siliconSigner = SiliconSigner(authContextCache: authContextCache, keyMetadataStore: keyMetadataStore)
    
    private lazy var siliconVerifier = SiliconVerifier(secItems: secItems, secKeys: secKeys, keyMetadataStore: keyMetadataStore)
    
    private lazy var siliconJose = SiliconJose(keyMetadataStore: keyMetadataStore, pubKeyData: pubKeyData)
    
    private lazy var siliconRandomGen = SiliconRandomGen(secRandom: secRandom)
    
    // MARK: - Definition
    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    public func definition() -> ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('ReactNativeSilicon')` in JavaScript.
        Name("ReactNativeSilicon")

        // ---- Device ----
        
        AsyncFunction("getCapabilities") { () -> [String: Any] in
            return siliconDevice.getCapabilities().toBridgeMap()
        }
          
        // ---- Keystore Manager ----
        
        AsyncFunction("generateKey") { (alias: String, opts: GenerateKeyOptions, attestChallenge: Data?) -> [String: Any] in
            // Uint8Array (TS) fails to map to Data correctly when it is nested inside the options
            // so we extract it out of the options on the TS side and then inject it back in here
            opts.attestChallenge = attestChallenge
            
            return await siliconKSM.generateKey(alias: alias, opts: opts).toBridgeMap()
        }
        
        AsyncFunction("deleteKey") { (alias: String) -> [String: Any] in
            return siliconKSM.deleteKey(alias: alias).toBridgeMap()
        }
        
        AsyncFunction("deleteAllKeys") { (prefix: String?) -> [String: Any] in
            return siliconKSM.deleteAllKeys(prefix: prefix).toBridgeMap()
        }
        
        AsyncFunction("keyExists") { (alias: String) -> [String: Any] in
            return siliconKSM.keyExists(alias: alias).toBridgeMap()
        }
        
        AsyncFunction("listKeys") { (prefix: String?) -> [String: Any] in
            return siliconKSM.listKeys(prefix: prefix).toBridgeMap()
        }
        
        AsyncFunction("getPubKey") { (alias: String, format: PubKeyFormat) -> [String: Any] in
            return siliconKSM.getPubKey(alias: alias, format: format).toBridgeMap()
        }
        
        AsyncFunction("attestKey") { (alias: String, format: AttestFormat, pubKeyFormat: PubKeyFormat) -> [String: Any] in
            return await siliconKSM.attestKey(alias: alias, format: format, pubKeyFormat: pubKeyFormat).toBridgeMap()
        }
        
        AsyncFunction("getKeyInfo") { (alias: String) -> [String: Any] in
            return siliconKSM.getKeyInfo(alias: alias).toBridgeMap()
        }
        
        AsyncFunction("validateKey") { (alias: String) -> [String: Any] in
            return siliconKSM.validateKey(alias: alias).toBridgeMap()
        }
        
        // ---- Sign/Verify ----
        
        AsyncFunction("sign") { (alias: String, payloadStr: String?, payloadByteArr: Data?, opts: SignOptions) -> [String: Any] in
            do {
                // Convert the payload to PayloadType
                let bridgePayload = BridgePayloadRecord()
                bridgePayload.text = payloadStr
                bridgePayload.bytes = payloadByteArr
                let payload = try bridgePayload.toPayloadType()
                
                return await siliconSigner.sign(alias: alias, payload: payload, opts: opts).toBridgeMap()
                
            } catch {
                return SiliconResult<Never>.failure(
                    code: .INTERNAL_ERROR,
                    message: error.localizedDescription,
                    nativeStack: nil
                ).toBridgeMap()
            }
        }
        
        AsyncFunction("verify") { (
            payloadStr: String?,
            payloadByteArr: Data?,
            signatureStr: String?,
            signatureByteArr: Data?,
            bridgeOpts: BridgeVerifyOptions
        ) -> [String: Any] in
            do {
                // Pack payload into PayloadType
                let bridgePayload = BridgePayloadRecord()
                bridgePayload.text = payloadStr
                bridgePayload.bytes = payloadByteArr
                let payload = try bridgePayload.toPayloadType()
                
                // Pack signature into PayloadType
                let bridgeSignature = BridgePayloadRecord()
                bridgeSignature.text = signatureStr
                bridgeSignature.bytes = signatureByteArr
                let signature = try bridgeSignature.toPayloadType()
                
                // Pack bridgeOpts.pubkeyStr and bridgeOpts.pubkeyBytes into PayloadType
                var pubkey: PayloadType?
                if bridgeOpts.pubkeyBytes != nil || bridgeOpts.pubkeyStr != nil {
                    let bridgePubkey = BridgePayloadRecord()
                    bridgePubkey.text = bridgeOpts.pubkeyStr
                    bridgePubkey.bytes = bridgeOpts.pubkeyBytes
                    pubkey = try bridgePubkey.toPayloadType()
                }
                let opts = VerifyOptions(
                    alias: bridgeOpts.alias,
                    pubkey: pubkey,
                    algorithm: bridgeOpts.algorithm
                )
                
                return siliconVerifier.verify(payload: payload, signature: signature, opts: opts).toBridgeMap()
                
            } catch {
                return SiliconResult<Never>.failure(
                    code: .INTERNAL_ERROR,
                    message: error.localizedDescription,
                    nativeStack: nil
                ).toBridgeMap()
            }
        }
        
        // ---- Random Generator ----
        
        AsyncFunction("generateSecureRandomBytes") { (length: Int, format: RandomBytesFormat) -> [String: Any] in
            return siliconRandomGen.generate(length: length, format: format).toBridgeMap()
        }

        // ---- JOSE ----
        
        AsyncFunction("getJwk") { (alias: String, digest: KeyDigests?) -> [String: Any] in
            return siliconJose.getJwk(alias: alias, digest: digest).toBridgeMap()
        }
    }
}
